Source/WebCore/platform/mediastream/gstreamer/RealtimeIncomingSourceGStreamer.cpp

```diff
@@ -179,12 +135,120 @@ std::optional<int> RealtimeIncomingSourceGStreamer::registerClient(GRefPtr<GstEl
     gst_app_sink_set_callbacks(GST_APP_SINK(sink), &callbacks, new ThreadSafeWeakPtr { *this }, [](void* data) {
         delete static_cast<ThreadSafeWeakPtr<RealtimeIncomingSourceGStreamer>*>(data);
     });
+}
+
+void RealtimeIncomingSourceGStreamer::configureFakeVideoSink(GstElement* sink)
+{
+    g_object_set(sink, "signal-handoffs", TRUE, nullptr);
+
+    auto handoffCallback = G_CALLBACK(+[](GstElement*, GstBuffer* buffer, GstPad* pad, gpointer userData) {
+        auto source = reinterpret_cast<ThreadSafeWeakPtr<RealtimeIncomingSourceGStreamer>*>(userData);
+        auto strongSource = source->get();
+        if (!strongSource)
+            return;
+
+        auto caps = adoptGRef(gst_pad_get_current_caps(pad));
+        auto sample = adoptGRef(gst_sample_new(buffer, caps.get(), nullptr, nullptr));
+        // dispatchSample might trigger RealtimeMediaSource::notifySettingsDidChangeObservers()
+        // which expects to run in the main thread.
+        callOnMainThread([source = ThreadSafeWeakPtr { *strongSource.get() }, sample = WTFMove(sample)]() mutable {
+            auto strongSource = source.get();
+            if (!strongSource)
+                return;
+
+            strongSource->dispatchSample(WTFMove(sample));
+        });
+    });
+    g_signal_connect_data(sink, "preroll-handoff", handoffCallback, new ThreadSafeWeakPtr { *this },
+        [](void* data, GClosure*) {
+            delete static_cast<ThreadSafeWeakPtr<RealtimeIncomingSourceGStreamer>*>(data);
+        }, static_cast<GConnectFlags>(0));
+    g_signal_connect_data(sink, "handoff", handoffCallback, new ThreadSafeWeakPtr { *this },
+        [](void* data, GClosure*) {
+            delete static_cast<ThreadSafeWeakPtr<RealtimeIncomingSourceGStreamer>*>(data);
+        }, static_cast<GConnectFlags>(0));
+
+    auto pad = adoptGRef(gst_element_get_static_pad(sink, "sink"));
+    gst_pad_add_probe(pad.get(), GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM, reinterpret_cast<GstPadProbeCallback>(+[](GstPad* pad, GstPadProbeInfo* info, gpointer userData) -> GstPadProbeReturn {
+        auto source = reinterpret_cast<ThreadSafeWeakPtr<RealtimeIncomingSourceGStreamer>*>(userData);
+        auto strongSource = source->get();
+        if (!strongSource)
+            return GST_PAD_PROBE_OK;
+        GRefPtr event = GST_PAD_PROBE_INFO_EVENT(info);
+        auto sink = adoptGRef(gst_pad_get_parent_element(pad));
+        strongSource->handleDownstreamEvent(sink.get(), WTFMove(event));
+        return GST_PAD_PROBE_OK;
+    }), new ThreadSafeWeakPtr { *this }, [](void* data) {
+        delete static_cast<ThreadSafeWeakPtr<RealtimeIncomingSourceGStreamer>*>(data);
+    });
+}
+
+void RealtimeIncomingSourceGStreamer::handleDownstreamEvent(GstElement* sink, GRefPtr<GstEvent>&& event)
+{
+    switch (GST_EVENT_TYPE(event.get())) {
+    case GST_EVENT_STREAM_START:
+    case GST_EVENT_CAPS:
+    case GST_EVENT_SEGMENT:
+    case GST_EVENT_STREAM_COLLECTION:
+        return;
```

Comment:
You don't need any of this if you have the default. And up to you, if you're just handling one case, it could be an if.

Reply:
I do, the default it to break, not return.